File: backend/app/models/achievement.py
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Any

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.models.user import User


class Achievement(Base):
    """Global achievement definitions, seeded via migration."""

    __tablename__ = "achievements"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    key: Mapped[str] = mapped_column(String(64), unique=True, index=True, nullable=False)
    category: Mapped[str] = mapped_column(String(32), nullable=False)  # consistency | strength | volume | secret
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(String(500), nullable=False)
    icon: Mapped[str] = mapped_column(String(32), nullable=False, default="emoji_events")
    criteria: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)
    is_secret: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False, server_default="false")


class UserAchievement(Base):
    __tablename__ = "user_achievements"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    achievement_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("achievements.id"), nullable=False)
    progress_value: Mapped[float] = mapped_column(Float, nullable=False, default=0)
    target_value: Mapped[float] = mapped_column(Float, nullable=False, default=1)
    unlocked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    user: Mapped["User"] = relationship(back_populates="achievements")
    achievement: Mapped["Achievement"] = relationship()

File: app/lib/features/achievements/domain/achievement.dart
class Achievement {
  const Achievement({
    required this.id,
    required this.key,
    required this.category,
    required this.title,
    required this.description,
    required this.icon,
    required this.isSecret,
  });

  final String id;
  final String key;
  final String category; // consistency | strength | volume | secret
  final String title;
  final String description;
  final String icon;
  final bool isSecret;

  factory Achievement.fromJson(Map<String, dynamic> json) => Achievement(
    id: json['id'] as String,
    key: json['key'] as String,
    category: json['category'] as String,
    title: json['title'] as String,
    description: json['description'] as String,
    icon: json['icon'] as String,
    isSecret: json['is_secret'] as bool,
  );
}

class UserAchievement {
  const UserAchievement({
    required this.id,
    required this.achievement,
    required this.progressValue,
    required this.targetValue,
    this.unlockedAt,
  });

  final String id;
  final Achievement achievement;
  final double progressValue;
  final double targetValue;
  final DateTime? unlockedAt;

  bool get isUnlocked => unlockedAt != null;
  double get progressFraction => targetValue <= 0 ? 0 : (progressValue / targetValue).clamp(0, 1);

  factory UserAchievement.fromJson(Map<String, dynamic> json) => UserAchievement(
    id: json['id'] as String,
    achievement: Achievement.fromJson(json['achievement'] as Map<String, dynamic>),
    progressValue: (json['progress_value'] as num).toDouble(),
    targetValue: (json['target_value'] as num).toDouble(),
    unlockedAt: json['unlocked_at'] != null ? DateTime.parse(json['unlocked_at'] as String) : null,
  );
}

class PersonalRecord {
  const PersonalRecord({
    required this.id,
    required this.exerciseId,
    required this.weightKg,
    required this.reps,
    required this.est1RmKg,
    required this.achievedAt,
  });

  final String id;
  final String exerciseId;
  final double weightKg;
  final int reps;
  final double est1RmKg;
  final DateTime achievedAt;

  factory PersonalRecord.fromJson(Map<String, dynamic> json) => PersonalRecord(
    id: json['id'] as String,
    exerciseId: json['exercise_id'] as String,
    weightKg: (json['weight_kg'] as num).toDouble(),
    reps: json['reps'] as int,
    est1RmKg: (json['est_1rm_kg'] as num).toDouble(),
    achievedAt: DateTime.parse(json['achieved_at'] as String),
  );
}
